### PR TITLE
Do not rely on exceptions in Struct#dig

### PR DESCRIPTION
**Pull request by kirs**
This attempts to solve https://github.com/oracle/truffleruby/issues/2299.

Not relying on exception flow shows much better perf in interpreter mode.

```ruby
require 'benchmark/ips'

puts RUBY_DESCRIPTION

struct = Struct.new(:foo).new(14)

Benchmark.ips do |x|
  x.report('found') do
      struct[:foo]
  end

  x.report('not-found') do
    struct.dig(:bar)
  end

  x.compare!
end
```

**Before**

```
$ jt ruby demo.rb
Using Interpreted TruffleRuby: mxbuild/truffleruby-jvm
$ /src/github.com/shopify/truffleruby/mxbuild/truffleruby-jvm/languages/ruby/bin/ruby \
  --experimental-options \
  --core-load-path=/src/github.com/shopify/truffleruby/src/main/ruby/truffleruby \
  demo.rb
[ruby] WARNING gem paths: /usr/local/bundle are not marked as installed by TruffleRuby. They might belong to another Ruby implementation and break unexpectedly. Configure your Ruby manager to use TruffleRuby, or `unset GEM_HOME GEM_PATH`. See https://github.com/oracle/truffleruby/blob/master/doc/user/ruby-managers.md
truffleruby 21.1.0-dev-2855d7c9, like ruby 2.7.2, Interpreted JVM [x86_64-linux]
Warming up --------------------------------------
               found    57.241k i/100ms
           not-found     5.873k i/100ms
Calculating -------------------------------------
               found    541.688k (±13.0%) i/s -      2.690M in   5.073909s
           not-found     56.472k (±13.2%) i/s -    281.904k in   5.079920s

Comparison:
               found:   541687.5 i/s
           not-found:    56472.2 i/s - 9.59x  (± 0.00) slower
```

**After**

```
$ jt ruby demo.rb
Using Interpreted TruffleRuby: mxbuild/truffleruby-jvm
$ /src/github.com/shopify/truffleruby/mxbuild/truffleruby-jvm/languages/ruby/bin/ruby \
  --experimental-options \
  --core-load-path=/src/github.com/shopify/truffleruby/src/main/ruby/truffleruby \
  demo.rb
[ruby] WARNING gem paths: /usr/local/bundle are not marked as installed by TruffleRuby. They might belong to another Ruby implementation and break unexpectedly. Configure your Ruby manager to use TruffleRuby, or `unset GEM_HOME GEM_PATH`. See https://github.com/oracle/truffleruby/blob/master/doc/user/ruby-managers.md
truffleruby 21.1.0-dev-2855d7c9, like ruby 2.7.2, Interpreted JVM [x86_64-linux]
Warming up --------------------------------------
               found    57.851k i/100ms
           not-found    52.936k i/100ms
Calculating -------------------------------------
               found    546.247k (± 3.8%) i/s -      2.777M in   5.091750s
           not-found    536.844k (± 2.0%) i/s -      2.700M in   5.030927s

Comparison:
               found:   546247.3 i/s
           not-found:   536844.3 i/s - same-ish: difference falls within error

```

### CODE REVIEW

File: src/main/ruby/truffleruby/core/struct.rb [hunk 1]
```diff
@@ -253,12 +253,8 @@ def initialize_copy(other)
...
 
   def dig(key, *more)
-    result = nil
-    begin
-      result = self[key]
-    rescue IndexError, NameError
-      nil # nothing found with key
-    end
+    return unless _attrs.include?(key.to_sym)
+    result = Primitive.object_hidden_var_get(self, key)
```
Comment (eregon) [inline src/main/ruby/truffleruby/core/struct.rb:257]: `Struct#dig` should also handle key being an Integer:
```ruby
ruby -e 'p Struct.new(:a,:b).new(:one, :two).dig(1)' # => 2
```

There is also a partial copy of this logic in `src/main/ruby/truffleruby/core/truffle/diggable.rb`.
To solve that I suggest adding a new private method `dig1` just below `[]` and then essentially copy-paste the logic of `[]` and adapt it in `dig1` (you'll also need a variant/copy of `check_index_var`). And simply call `dig1` in `dig`.

Comment (eregon) [reply src/main/ruby/truffleruby/core/struct.rb:257]: `dig1` is meant as "dig with 1 index/dig 1 time", but could also be `read_or_nil` or so to be more explicit.

Comment (kirs) [reply src/main/ruby/truffleruby/core/struct.rb]: Thanks for feedback!

I rewrote it with `read_or_nil`. I found it hard to make it reused in other places (like `def check_index_var`) so I inlined most of the logic.

I also added a regression test for the edge case you've described.

Comment (chrisseaton) [reply src/main/ruby/truffleruby/core/struct.rb]: I tried to do a pass to inline `check_index_var` myself, and yes it's hard to factor out because of the exception and their messages. You'd need error codes or something complex like that. Probably best not to.

**Comment by chrisseaton** [comment]
Thanks for this work. 10x slower for missing is another example of a 'performance cliff'. It's hard to explain to users why a missing field is an order of magnitude worse when it seems to be conceptually the same complexity.

File: src/main/ruby/truffleruby/core/struct.rb (hunk 1 above)
Comment (eregon) [reply src/main/ruby/truffleruby/core/struct.rb]: Yes, I was thinking an inlined copy of `check_index_var` and adapting as needed makes sense since there is a lot of differences. Sorry I wasn't clear.

**Review by eregon — APPROVED** [review]
Awesome, thank you for the PR!
